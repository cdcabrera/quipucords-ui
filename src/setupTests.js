import { configure, mount, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { act } from 'react-dom/test-utils';
configure({ adapter: new Adapter() });

/**
 * Emulate for component checks
 */
jest.mock('i18next', () => {
  const Test = function () { // eslint-disable-line
    this.use = () => this;
    this.init = () => Promise.resolve();
    this.changeLanguage = () => Promise.resolve();
  };
  return new Test();
});

/**
 * Enzyme for components using hooks.
 *
 * @param {React.ReactNode} component
 * @param {object} options
 * @param {Function} options.callback
 * @param {object} options.options
 *
 * @returns {Promise<null>}
 */
global.mountHookComponent = async (component, { callback, ...options } = {}) => {
  let mountedComponent = null;
  await act(async () => {
    mountedComponent = mount(component, options);
  });
  mountedComponent?.update();

  if (typeof callback === 'function') {
    await act(async () => {
      await callback({ component: mountedComponent });
    });
    mountedComponent?.update();
  }

  return mountedComponent;
};

global.mountHookWrapper = global.mountHookComponent;

/**
 * Enzyme for components using hooks.
 *
 * @param {React.ReactNode} component
 * @param {object} options
 * @param {Function} options.callback
 * @param {object} options.options
 *
 * @returns {Promise<null>}
 */
global.shallowHookComponent = async (component, { callback, ...options } = {}) => {
  let mountedComponent = null;
  await act(async () => {
    mountedComponent = shallow(component, options);
  });
  mountedComponent?.update();

  if (typeof callback === 'function') {
    await act(async () => {
      await callback({ component: mountedComponent });
    });
    mountedComponent?.update();
  }

  return mountedComponent;
};

global.shallowHookWrapper = global.shallowHookComponent;

/*
 * This is a temporary patch for applying a global Jest "beforeAll", based on
 * jest-prop-type-error, https://www.npmjs.com/package/jest-prop-type-error
 */
beforeAll(() => {
  const { error } = console;

  console.error = (message, ...args) => {
    if (/(Invalid prop|Failed prop type)/gi.test(message)) {
      throw new Error(message);
    }

    error.apply(console, [message, ...args]);
  };
});
