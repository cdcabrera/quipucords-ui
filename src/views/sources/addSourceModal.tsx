/**
 * Add Source Modal Component
 *
 * This component displays a modal for adding or editing a source of a specific type. It provides
 * a form to input source details including name, hosts, port, credential, and SSL settings.
 *
 * @module addSourceModal
 */
import React, { FormEvent, useEffect, useState } from 'react';
import {
  ActionGroup,
  Button,
  Checkbox,
  Form,
  FormContextProvider,
  FormGroup,
  FormSelect,
  FormSelectOption,
  HelperText,
  Modal,
  ModalVariant,
  Radio,
  TextArea,
  TextInput
} from '@patternfly/react-core';
import { SimpleDropdown } from '../../components/simpleDropdown/simpleDropdown';
import { TypeaheadCheckboxes } from '../../components/typeAheadCheckboxes/typeaheadCheckboxes';
import { helpers } from '../../helpers';
import { useGetCredentialsApi } from '../../hooks/useCredentialApi';
import { type SourceType } from '../../types/types';

interface AddSourceModalProps {
  isOpen: boolean;
  source?: SourceType;
  sourceType?: string;
  onClose?: () => void;
  onSubmit?: (payload: any) => void;
  useGetCredentials?: typeof useGetCredentialsApi;
}

interface SourceFormFieldsProps {
  formData: any;
  credOptions: { value: string; label: string }[];
  isNetwork: boolean;
  handleInputChange: (field: string, value: any) => void;
  setValue: (field: string, value: any) => void;
  getValue: (field: string) => any;
}

const useSourceForm = (sourceType: string | undefined, source?: SourceType) => {
  const [formData, setFormData] = useState({
    credentials: source?.credentials?.map(c => c.id) || [],
    useParamiko: source?.options?.use_paramiko ?? false,
    sslVerify: source?.options?.ssl_cert_verify ?? true,
    sslProtocol: source?.options?.disable_ssl ? 'Disable SSL' : source?.options?.ssl_protocol || 'SSLv23',
    name: source?.name || '',
    hosts: source?.hosts?.join(',') || '',
    port: source?.port ? String(source.port) : ''
  });

  const sourceTypeValue = source?.source_type || sourceType?.split(' ')?.shift()?.toLowerCase();
  const isNetwork = sourceTypeValue === 'network';

  const resetForm = () => {
    setFormData({
      credentials: [],
      useParamiko: false,
      sslVerify: true,
      sslProtocol: 'SSLv23',
      name: '',
      hosts: '',
      port: ''
    });
  };

  const handleInputChange = (field: string, value: any) => {
    setFormData(prev => ({ ...prev, [field]: value }));
  };

  const filterFormData = (values: any) => {
    const { credentials, useParamiko, sslVerify, sslProtocol, name, hosts, port } = formData;
    const payload = {
      name: name,
      credentials: credentials.map(c => Number(c)),
      hosts: hosts.split(','),
      port: port || (isNetwork ? '22' : '443'),
      options: !isNetwork
        ? {
            ssl_cert_verify: sslProtocol !== 'Disable SSL' && sslVerify,
            disable_ssl: sslProtocol === 'Disable SSL',
            ...(sslProtocol !== 'Disable SSL' && { ssl_protocol: sslProtocol })
          }
        : {
            use_paramiko: useParamiko
          },
      ...(!source && { source_type: sourceTypeValue }),
      ...(source && { id: source.id })
    };

    return payload;
  };

  return {
    formData,
    isNetwork,
    resetForm,
    handleInputChange,
    filterFormData
  };
};

const SourceFormFields: React.FC<SourceFormFieldsProps> = ({
  formData,
  credOptions,
  isNetwork,
  handleInputChange,
  setValue,
  getValue
}) => (
  <>
    <FormGroup label="Name" isRequired fieldId="name">
      <TextInput
        value={formData?.name}
        placeholder="Enter a name for the credential"
        isRequired
        type="text"
        id="credential-name"
        name="name"
        onChange={event => handleInputChange('name', (event.target as HTMLInputElement).value)}
      />
    </FormGroup>

    <FormGroup label="Credentials" fieldId="credentials" isRequired>
      {isNetwork ? (
        <TypeaheadCheckboxes
          onChange={(selections: string[]) => {
            const selectedIds = selections.map(Number);
            const validIds = selectedIds.filter(id => !isNaN(id));
            handleInputChange('credentials', validIds);
          }}
          options={credOptions}
          selectedOptions={formData.credentials?.map(String) || []}
        />
      ) : (
        <FormSelect
          value={formData.credentials[0]?.toString() || ''}
          onChange={(event: FormEvent<HTMLSelectElement>, value: string) => {
            const selectedCredential = Number(value);
            if (!isNaN(selectedCredential)) {
              handleInputChange('credentials', [selectedCredential]);
            } else {
              handleInputChange('credentials', []);
            }
          }}
        >
          <FormSelectOption value="" label="Select a credential" isDisabled />
          {credOptions.map(option => (
            <FormSelectOption
              key={option.value}
              value={option.value.toString()}
              label={option.label}
            />
          ))}
        </FormSelect>
      )}
    </FormGroup>

    {isNetwork ? (
      <>
        <FormGroup label="Search addresses" isRequired fieldId="hosts">
          <TextArea
            placeholder="Enter values separated by commas"
            value={getValue('hosts')}
            onChange={(_ev, val) => setValue('hosts', val)}
            isRequired
            id="source-hosts"
            name="hosts"
          />
          <HelperText>
            Type IP addresses, IP ranges, and DNS host names. Wildcards are valid. Use CIDR or Ansible notation for ranges.
          </HelperText>
        </FormGroup>

        <FormGroup label="Port" fieldId="port">
          <TextInput
            value={getValue('port')}
            placeholder="Optional"
            type="text"
            id="source-port"
            name="port"
            onChange={ev => {
              setValue('port', (ev.target as HTMLInputElement).value);
            }}
          />
          <HelperText>Default port is 22</HelperText>
        </FormGroup>
      </>
    ) : (
      <>
        <FormGroup label="IP address or hostname" isRequired fieldId="hosts">
          <TextInput
            value={getValue('hosts')}
            onChange={(_ev, val) => setValue('hosts', val)}
            isRequired
            id="source-hosts"
            name="hosts"
          />
          <HelperText>Enter an IP address or hostname</HelperText>
        </FormGroup>

        <FormGroup label="Port" fieldId="port">
          <TextInput
            value={getValue('port')}
            placeholder="Optional"
            type="text"
            id="source-port"
            name="port"
            onChange={ev => {
              setValue('port', (ev.target as HTMLInputElement).value);
            }}
          />
          <HelperText>Default port is 443</HelperText>
        </FormGroup>
      </>
    )}

    {isNetwork ? (
      <FormGroup fieldId="paramiko">
        <Checkbox
          key="paramiko"
          label="Connect using Paramiko instead of Open SSH"
          id="paramiko"
          isChecked={formData.useParamiko}
          onChange={(_ev, checked) => handleInputChange('useParamiko', checked)}
        />
      </FormGroup>
    ) : (
      <>
        <FormGroup label="Connection" fieldId="connection">
          <SimpleDropdown
            isFullWidth
            label={formData.sslProtocol}
            variant="default"
            onSelect={item => handleInputChange('sslProtocol', item)}
            dropdownItems={['SSLv23', 'TLSv1', 'TLSv1.1', 'TLSv1.2', 'Disable SSL']}
          />
        </FormGroup>

        <FormGroup fieldId="ssl_verify">
          <Checkbox
            key="ssl_verify"
            label="Verify SSL certificate"
            id="ssl_verify"
            isDisabled={formData.sslProtocol === 'Disable SSL'}
            isChecked={formData.sslProtocol !== 'Disable SSL' && formData.sslVerify}
            onChange={(_ev, checked) => handleInputChange('sslVerify', checked)}
          />
        </FormGroup>
      </>
    )}
  </>
);

const AddSourceModal: React.FC<AddSourceModalProps> = ({
  isOpen,
  source,
  sourceType,
  onClose = Function.prototype,
  onSubmit = Function.prototype,
  useGetCredentials = useGetCredentialsApi
}) => {
  const { getCredentials } = useGetCredentials();
  const [credOptions, setCredOptions] = useState<{ value: string; label: string }[] | []>([]);

  const {
    formData,
    isNetwork,
    handleInputChange,
    resetForm,
    filterFormData
  } = useSourceForm(sourceType, source);

  useEffect(() => {
    if (isOpen) {
      getCredentials({
        params: {
          cred_type: sourceType?.split(' ')?.shift()?.toLowerCase() || ''
        }
      })
        .then(response => {
          const updatedOptions = response?.data?.results?.map(({ name, id }) => ({ label: name, value: `${id}` }));
          setCredOptions(updatedOptions || []);
        })
        .catch(err => {
          if (!helpers.TEST_MODE) {
            console.error(err);
          }
        });
    }
  }, [getCredentials, sourceType, isOpen]);

  useEffect(() => {
    if (!isOpen) {
      resetForm();
    }
  }, [isOpen, sourceType, resetForm]);

  const onAdd = (values: any) => {
    const filteredData = filterFormData(values);
    console.log("!!!!!!!!");  // Check if name is correctly set here

    console.log(filteredData.name);  // Check if name is correctly set here
    onSubmit(filteredData);
  };

  return (
    <Modal
      variant={ModalVariant.small}
      title={(source && 'Edit') || `Add Source: ${sourceType}`}
      isOpen={isOpen}
      onClose={() => {
        resetForm();
        onClose();
      }}
    >
      <FormContextProvider
        initialValues={{
          name: formData.name,
          hosts: formData.hosts,
          port: formData.port
        }}
      >
        {({ setValue, getValue, values }) => (
          <Form>
            <SourceFormFields
              formData={formData}
              credOptions={credOptions}
              isNetwork={isNetwork}
              handleInputChange={handleInputChange}
              setValue={setValue}
              getValue={getValue}
            />
            <ActionGroup>
              <Button
                variant="primary"
                onClick={() => {
                  onAdd(values);
                  resetForm();
                  onClose();
                }}
              >
                Save
              </Button>
              <Button variant="link" onClick={() => onClose()}>
                Cancel
              </Button>
            </ActionGroup>
          </Form>
        )}
      </FormContextProvider>
    </Modal>
  );
};

export { AddSourceModal as default, AddSourceModal, type AddSourceModalProps };
