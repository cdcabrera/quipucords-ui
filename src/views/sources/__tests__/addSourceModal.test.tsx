import React, { act } from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { shallowComponent } from '../../../../config/jest.setupTests';
import { AddSourceModal } from '../addSourceModal';
import axios from 'axios';

describe('AddSourceModal', () => {
  let mockOnClose;
  let mockOnSubmit;

  //count?: number;
  //   next?: string | null;
  //   previous?: string | null;
  //   results?: CredentialType[];

  beforeEach(async () => {
    jest.spyOn(axios, 'get').mockImplementation(() => Promise.resolve({}));

    await act(async () => {
      mockOnClose = jest.fn();
      mockOnSubmit = jest.fn();
      const mockUseCredentialsApi = () => ({
        callbackSuccess: () => jest.fn(),
        callbackError: () => jest.fn(),
        apiCall: () => Promise.resolve({}),
        getCredentials: () => Promise.resolve({ data: { results: [] } })
      });
      await render(
        <AddSourceModal
          isOpen={true}
          sourceType="network"
          source={undefined}
          onClose={mockOnClose}
          onSubmit={mockOnSubmit}
          // useGetCredentials={mockUseCredentialsApi}
        />
      );
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a basic component', async () => {
    const component = await shallowComponent(<AddSourceModal isOpen={true} />);
    expect(component).toMatchSnapshot('basic');
  });

  it('should have the correct title', () => {
    const title = screen.getByText(/Add\sSource:\snetwork/i);
    expect(title).toMatchSnapshot('title');
  });

  it('should call onSubmit with the correct filtered data when "Save" is clicked', async () => {
    const user = userEvent.setup();
    await user.type(screen.getByPlaceholderText('Enter a name for the source'), 'Test Source');
    await user.click(screen.getByText('Save'));

    expect(mockOnSubmit.mock.calls).toMatchSnapshot('onSubmit, filtered data');
  });

  it('should call onClose', async () => {
    const user = userEvent.setup();
    await user.click(screen.getByText('Cancel'));

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });
});
