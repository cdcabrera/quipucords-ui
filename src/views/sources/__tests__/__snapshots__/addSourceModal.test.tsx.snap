// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddSourceModal should call onSubmit with the correct filtered data when "Save" is clicked: onSubmit, filtered data 1`] = `
[
  [
    {
      "credentials": [],
      "hosts": [
        "",
      ],
      "name": "Test Source",
      "options": {
        "use_paramiko": false,
      },
      "port": "22",
      "source_type": "network",
    },
  ],
]
`;

exports[`AddSourceModal should have the correct title: title 1`] = `
<span
  class="pf-v5-c-modal-box__title-text"
>
  Add source: network
</span>
`;

exports[`AddSourceModal should render a basic component: basic 1`] = `
<Modal
  actions={[]}
  appendTo={[Function]}
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={true}
  onClose={[Function]}
  ouiaSafe={true}
  position="default"
  showClose={true}
  title="Add source: undefined"
  titleIconVariant={null}
  titleLabel=""
  variant="small"
>
  <FormContextProvider
    initialValues={
      {
        "hosts": "",
        "name": "",
        "port": "",
      }
    }
  >
    [Function]
  </FormContextProvider>
</Modal>
`;
